File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/service/impl/CollectCategoryServiceImpl.java
package com.windea.demo.cloudcollect.core.service.impl;

import com.windea.demo.cloudcollect.core.domain.entity.*;
import com.windea.demo.cloudcollect.core.exception.NotFoundException;
import com.windea.demo.cloudcollect.core.repository.CollectCategoryRepository;
import com.windea.demo.cloudcollect.core.repository.CollectRepository;
import com.windea.demo.cloudcollect.core.service.CollectCategoryService;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;

@Service
public class CollectCategoryServiceImpl implements CollectCategoryService {
	private final CollectCategoryRepository repository;
	private final CollectRepository collectRepository;

	public CollectCategoryServiceImpl(CollectCategoryRepository repository, CollectRepository collectRepository) {
		this.repository = repository;
		this.collectRepository = collectRepository;
	}


	@Transactional
	@Override
	public void create(CollectCategory category, User user) {
		category.setUser(user);
		repository.save(category);
	}

	@Transactional
	@Override
	public void delete(Long id) {
		repository.deleteById(id);
	}

	@Transactional
	@Override
	public void modify(Long id, CollectCategory category) {
		var rawCategory = repository.getOne(id);
		rawCategory.setName(category.getName());
		rawCategory.setSummary(category.getSummary());
		repository.save(rawCategory);
	}

	@Cacheable("collectCategory")
	@Override
	public CollectCategory get(Long id) {
		return repository.findById(id).orElseThrow(NotFoundException::new);
	}

	@Cacheable("collectCategory.collectPage")
	@Override
	public Page<Collect> getCollectPage(Long id, Pageable pageable) {
		return collectRepository.findByCategory_IdAndDeleteStatusFalse(id, pageable);
	}

	@Cacheable("collectCategory.collectCount")
	@Override
	public Long getCollectCount(Long id) {
		return collectRepository.countByCategory_IdAndDeleteStatusFalse(id);
	}

	@Cacheable("collectCategoryPage")
	@Override
	public Page<CollectCategory> findAll(Pageable pageable) {
		return repository.findAll(pageable);
	}

	@Cacheable("collectCategoryPage.byUser")
	@Override
	public Page<CollectCategory> findByUser(Long userId, Pageable pageable) {
		return repository.findByUser_Id(userId, pageable);
	}

	@Cacheable("collectCategoryPage.byUserAndName")
	@Override
	public Page<CollectCategory> findByUserAndName(Long userId, String name, Pageable pageable) {
		return repository.findByUser_IdAndNameContains(userId, name, pageable);
	}

	@Override
	public boolean exists(CollectCategory category) {
		var userId = category.getUser().getId();
		var name = category.getName();
		return repository.existsByUser_IdAndName(userId, name);
	}
}

File: cloud-collect-core/src/main/kotlin/com/windea/demo/cloudcollect/core/service/impl/CollectCategoryServiceImpl.kt
package com.windea.demo.cloudcollect.core.service.impl

import com.windea.demo.cloudcollect.core.domain.entity.*
import com.windea.demo.cloudcollect.core.exceptions.*
import com.windea.demo.cloudcollect.core.repository.*
import com.windea.demo.cloudcollect.core.service.*
import org.springframework.cache.annotation.*
import org.springframework.data.domain.*
import org.springframework.data.repository.*
import org.springframework.stereotype.*
import javax.transaction.*

@Service
@CacheConfig(cacheNames = ["collectCategory"])
class CollectCategoryServiceImpl(
	private val collectRepository: CollectRepository,
	private val categoryRepository: CollectCategoryRepository
) : CollectCategoryService {
	@Transactional
	@CacheEvict(allEntries = true)
	override fun create(category: CollectCategory, user: User): CollectCategory {
		val newCategory = category.copy(
			user = user
		)
		return categoryRepository.save(newCategory)
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun modify(category: CollectCategory) {
		categoryRepository.save(category)
	}
	
	@Transactional
	@CacheEvict(allEntries = true)
	override fun deleteById(id: Long) {
		categoryRepository.deleteById(id)
	}
	
	@Cacheable(key = "methodName + args")
	override fun findById(id: Long): CollectCategory {
		return categoryRepository.findByIdOrNull(id)?.lateInit() ?: throw NotFoundException()
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAll(pageable: Pageable): Page<CollectCategory> {
		return categoryRepository.findAll(pageable).map { it.lateInit() }
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByNameContainsAndUserId(userId: Long, name: String, pageable: Pageable): Page<CollectCategory> {
		return categoryRepository.findAllByNameContainsAndUserId(name, userId, pageable).map { it.lateInit() }
	}
	
	@Cacheable(key = "methodName + args")
	override fun findAllByUserId(userId: Long, pageable: Pageable): Page<CollectCategory> {
		return categoryRepository.findAllByUserId(userId, pageable).map { it.lateInit() }
	}
	
	override fun existsByNameAndUserId(name: String, userId: Long): Boolean {
		return categoryRepository.existsByNameAndUserId(name, userId)
	}
	
	private fun CollectCategory.lateInit() = this.apply {
		collectCount = collectRepository.countByCategoryId(this.id!!)
	}
	
	
	@Cacheable(key = "methodName + args")
	override fun getCollectPage(id: Long, pageable: Pageable): Page<Collect> {
		return collectRepository.findAllByCategoryId(id, pageable)
	}
}
